druntime/src/rt/aaA.d

```diff
@@ -210,182 +215,58 @@ Bucket[] allocBuckets(size_t dim) @trusted pure nothrow
 
 private void* allocEntry(scope const Impl* aa, scope const void* pkey)
 {
-    import rt.lifetime : _d_newitemU;
     import core.stdc.string : memcpy, memset;
 
-    immutable akeysz = aa.valoff;
-    void* res = void;
-    if (aa.entryTI)
-        res = _d_newitemU(aa.entryTI);
-    else
+    immutable akeysz = aa.keysz;
+    immutable allocsz = aa.valoff + aa.valsz;
+    auto flags = (aa.flags & Impl.Flags.hasPointers) ? 0 : GC.BlkAttr.NO_SCAN;
+    if (aa.flags & Impl.Flags.hasDtors)
     {
-        auto flags = (aa.flags & Impl.Flags.hasPointers) ? 0 : GC.BlkAttr.NO_SCAN;
-        res = GC.malloc(akeysz + aa.valsz, flags);
+        assert(aa.entryTI !is null);
+        flags |= GC.BlkAttr.FINALIZE;
     }
+    auto res = GC.malloc(allocsz, flags, aa.entryTI);
 
-    memcpy(res, pkey, aa.keysz); // copy key
-    memset(res + akeysz, 0, aa.valsz); // zero value
+    memcpy(res, pkey, akeysz); // copy key
+    memset(res + akeysz, 0, allocsz - akeysz); // zero everything after
```

Comment:
Note, I changed this to memset also the gap between key and value, because I've seen what problems that can cause, and it's probably at most 15 extra bytes to memset with the same call.